lib/core/src/Cardano/Wallet/Primitive/AddressDerivation/Sequential.hs

```diff
@@ -350,6 +366,116 @@ deserializeXPrvSeq (k, h) = (,)
+    getSpendingKey (Address bytes)
+        | let l = BS.length bytes in l == addrLenSingle || l == addrLenGrouped =
+            BS.take (addrLenSingle - 1) $ BS.drop 1 bytes
+        | otherwise =
+            error "InspectAddress: tried to inspect an incompatible address"
+
+    type DelegationKey SeqKey = ByteString
+    getDelegationKey (Address bytes)
+        | BS.length bytes == addrLenSingle =
+            Nothing
+
+        | BS.length bytes == addrLenGrouped =
+            Just $ BS.drop addrLenSingle bytes
+
+        | otherwise =
+            error "InspectAddress: tried to inspect an incompatible address"
+
+instance KeyToAddress 'Mainnet SeqKey where
+    keyToAddress = gKeyToAddress @'Mainnet . getKey
+
+instance KeyToAddress 'Testnet SeqKey where
+    keyToAddress = gKeyToAddress @'Testnet . getKey
+
+-- | We use this to define both instance separately to avoid having to carry a
+-- 'KnownNetwork' constraint around.
+gKeyToAddress :: forall n. KnownNetwork n => XPub -> Address
+gKeyToAddress key =
+    Address $ invariantSize addrLenSingle $ BL.toStrict $ runPut $ do
+        putWord8 (single @n)
+        putByteString $ xpubPublicKey key
+  where
+    invariantSize :: HasCallStack => Int -> ByteString -> ByteString
+    invariantSize len bytes
+        | BS.length bytes == len = bytes
+        | otherwise = error
+            $ "length was "
+            ++ show (BS.length bytes)
+            ++ ", but expected to be "
+            ++ (show len)
+
+-- Serialized length in bytes of a Single Address
+addrLenSingle :: Int
+addrLenSingle = 33
+
+-- Serialized length in bytes of a Grouped Address
+addrLenGrouped :: Int
+addrLenGrouped = 65
+
+-- | Verify the structure of a payload decoded from a Bech32 text string
+decodeShelleyAddress
+    :: forall n. KnownNetwork n
+    => ByteString
+    -> Either TextDecodingError Address
+decodeShelleyAddress bytes = do
+    case BS.length bytes of
+        n | n == singleAddressLength ->
+            when (BS.take 1 bytes /= BS.pack [single @n]) $
+                Left (invalidFirstByte (single @n))
+        n | n == groupedAddressLength ->
+            when (BS.take 1 bytes /= BS.pack [grouped @n]) $
+                Left (invalidFirstByte (grouped @n))
+        _ ->
+            Left $ TextDecodingError $
+                "Invalid Address length (" <> show (BS.length bytes)
+                <> "): expected either "
+                <> show singleAddressLength
+                <> " or "
+                <> show groupedAddressLength
+                <> " bytes."
+    return (Address bytes)
+  where
+    singleAddressLength = 33
+    groupedAddressLength = 65
+    invalidFirstByte discriminant = TextDecodingError
+        $ "Invalid Address first byte: "
+        <> B8.unpack (BS.take 1 bytes)
+        <> " =/= "
+        <> B8.unpack (BS.pack [discriminant])
+        <> "."
+
+-- | Embed some constants into a network type.
+class KnownNetwork (n :: Network) where
```

Comment:
I wonder if this is always the same irrespective of era ...

Reply:
Note that I'll rename `AddressDerivation.Sequential` to `AddressDerivation.Shelley` in the end to better clarify the purpose of this module. You're right that this really depends on the era, and that's why the division here shouldn't be on the discovery scheme but on the era!